sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala

```diff
@@ -621,6 +621,33 @@ object DataSourceStrategy
             right, translatedFilterToExpr, nestedPredicatePushdownEnabled)
         } yield sources.Or(leftFilter, rightFilter)
 
+      case notNull @ expressions.IsNotNull(_: AttributeReference) =>
+        // Not null filters on attribute references can always be pushed, also for collated columns.
+        val filter =
+          translateLeafNodeFilter(notNull, PushableColumn(nestedPredicatePushdownEnabled))
+        if (filter.isDefined && translatedFilterToExpr.isDefined) {
+          translatedFilterToExpr.get(filter.get) = predicate
+        }
+        filter
+
+      case isNull @ expressions.IsNull(_: AttributeReference) =>
+        // Is null filters on attribute references can always be pushed, also for collated columns.
+        val filter =
+          translateLeafNodeFilter(isNull, PushableColumn(nestedPredicatePushdownEnabled))
+        if (filter.isDefined && translatedFilterToExpr.isDefined) {
+          translatedFilterToExpr.get(filter.get) = predicate
+        }
+        filter
+
+      case p if p.references.exists(ref => SchemaUtils.hasNonUTF8BinaryCollation(ref.dataType)) =>
```

Comment:
should this be `case p if !p.isInstanceOf[expressions.IsNotNull] && !p.isInstanceOf[expressions.IsNull] ...`

Reply:
hmmm, what if it's `Not(Not(IsNotNull(...)))`? I feel checking references at this layer is risky. Shall we do it in `translateLeafNodeFilter`?

Reply:
Just checking for the IsNull/IsNotNull types is not enough. It would not cover `IsNotNull(GreaterThan(<collated_col>, x))`. So it is important that IsNotNull / IsNotNull is directly wrapped around an attribute reference.
We can actually only do this predicate widening at the top level, particularly because of the `Not`s. If we for example have `Not(EqualTo(<collated_col>, x))` we would end up with `Not(AlwaysTrue)` if we do the transformation in `translateLeafNodeFilter`, and that is incorrect